### PR TITLE
Improve the interoperability of all ObjC swizzling activities in the framework.

### LINKED ISSUE #3432: NSButton inside NSStackView does not emit state
After upgrade to 5.0.2, NSButton embedded inside NSStackView stopped emitting state signals. 

I will make pull request with test case.

Xcode version: 8.3
macOS version: 10.12.4

### PR DESCRIPTION
**Pull request by andersio**
Fixes #3432 and #3433.

Since fixing one with new test cases leads to another, it all ends up in the same branch. 😅

#### Highlights

1. https://github.com/ReactiveCocoa/ReactiveCocoa/pull/3435/commits/e8bfbfc40fade886f0458d07dfd7c45f7ab9cf11: Fixed an interoperability issue in the swizzling logic. The original implementation had not taken account of the existence of other swizzling users.

1. https://github.com/ReactiveCocoa/ReactiveCocoa/pull/3435/commits/b774210cba5c0579693c301ae95f5aa2fea7bbfa: Replaced the unreliable use of `#function` as keys with the setter selector, which is constant for a given proxy.

1. https://github.com/ReactiveCocoa/ReactiveCocoa/pull/3435/commits/d3b850e3c58302f796627d9ef38b0681e910f39b: Refined the DelegateProxy implementation, so that it always respects the latest implementation in the perceived class (in case it has been swizzled externally).

1. https://github.com/ReactiveCocoa/ReactiveCocoa/pull/3435/commits/434227c25add109440397e8acf1442ab18d1d063: `ActionProxy` now guards for the absence of `NSApp`. Probably only relevant in tests.

### CODE REVIEW

**Comment by andersio** [comment]
There is still an interoperability issue with the interception API of our own.

**Comment by andersio** [comment]
There is one corner case of interoperability that hasn't been fixed (https://github.com/ReactiveCocoa/ReactiveCocoa/pull/3435/commits/c35e9bab3ffedc8f6c097e338833bdffe203c2ab), because it could require non-trivial changes to the method interception logic.

I wonder if it worths the effort, since it is about supporting the simultaneous use of KVO/method interception and the proxies on the same setter... Is this a real world use case, heh?

/cc @ReactiveCocoa/reactivecocoa

**Comment by andersio** [comment]
Seems like the MapViewSpec keeps failing with Xcode 8.3. 🙃

File: ReactiveCocoa/AppKit/ActionProxy.swift [hunk 1]
```diff
@@ -2,7 +2,7 @@ import AppKit
 import ReactiveSwift
 import enum Result.NoError
 
-internal final class ActionProxy<Owner: AnyObject> {
+internal final class ActionProxy<Owner: AnyObject>: NSObject {
```
Comment (andersio) [inline ReactiveCocoa/AppKit/ActionProxy.swift]: The `NSObject` is added for KVO interoperability tests to pass. I expect this to be removed by supporting native Swift object in the KVO API.

File: ReactiveCocoa/DelegateProxy.swift [hunk 2]
```diff
@@ -53,36 +54,46 @@ extension DelegateProxy {
...
 			instance.swizzle((setter, newSetterImpl), key: hasSwizzledKey)
 
-			// Set the proxy as the delegate.
-			let realClass: AnyClass = class_getSuperclass(object_getClass(instance))
+			// Retrieve the original delegate setter.
+			//
+			// `instance.objcClass` (aka `+class`) would report the perceived
+			// class instead of the real runtime subclass generated by RAC or
+			// KVO.
+			let realClass: AnyClass = instance.objcClass
```
Comment (andersio) [inline ReactiveCocoa/DelegateProxy.swift]: While in most cases the superclass of the real runtime class is the original/perceived class, it is not true if the instance is first swizzled by RAC then by KVO. It would end up with being `NSKVONotifying_*_RACSwift`, i.e. two levels of inheritance.

**Comment by andersio** [comment]
(Tackled a couple of flaws and consolidated the patches.)

File: ReactiveCocoa/DelegateProxy.swift [hunk 3]
```diff
@@ -53,57 +54,64 @@ extension DelegateProxy {
...
+				let impl = class_getMethodImplementation(object.objcClass, selector)
+				unsafeBitCast(impl, to: Setter.self)(object, selector, delegate)
+			}
+
 			let newSetterImpl: @convention(block) (NSObject, AnyObject?) -> Void = { object, delegate in
-				let proxy = object.associations.value(forKey: key)!
-				proxy.forwardee = (delegate as! Delegate?)
+				let proxyMap = object.associations.value(forKey: proxyMapKey)
+
+				if let proxy = proxyMap.first(where: { $0.0 == setter })?.1 as! DelegateProxy<Delegate>? {
```
Comment (mdiep) [inline ReactiveCocoa/DelegateProxy.swift]: Why is `proxyMap` an array if we're only going to handle 1 item?

Comment (andersio) [inline ReactiveCocoa/DelegateProxy.swift]: In rare circumstances (e.g. collection views) there could be more than one delegate being proxied. After it being eventually public, users might also have classes that got multiple delegates. So it is made to support multiple proxies.

**Comment by andersio** [comment]
P.S. I wonder if we should drop a note in README or release note to discourage a mixed use of RAC and ReactiveObjC swizzling-fuelled API. There is no test coverage in this area (yet), and while they can live together, they gonna collide when being applied to the same selector. The scope of interoperability addressed here is limited to KVO and RAC itself.

File: ReactiveCocoa/DelegateProxy.swift (hunk 3 above)
Comment (mdiep) [inline ReactiveCocoa/DelegateProxy.swift]: I'll still don't understand. How could more than one be proxied? I don't see why a collection view would have more than one delegate.

Comment (andersio) [inline ReactiveCocoa/DelegateProxy.swift]: `dataSource` and `collectionViewLayout` are delegates, aren't they?

Comment (mdiep) [inline ReactiveCocoa/DelegateProxy.swift]: Okay, I understand what's happening now.

It seems like we could instead create an `AssociationKey` from `setter` since it's a `Selector`.

If for some reason that doesn't work, we should be able to use `[Selector: AnyObject]` since `Selector` is `Hashable`.

Comment (andersio) [inline ReactiveCocoa/DelegateProxy.swift]: These all work, but just a matter of preferences. The differences are minimal in the unit of bytes. 🙂

> It seems like we could instead create an AssociationKey from setter since it's a Selector.

It could collide with others if we use it directly. In the meantime, I'd prefer to avoid a new selector unless it is necessary.

> If for some reason that doesn't work, we should be able to use [Selector: AnyObject] since Selector is Hashable.

A dictionary is bulkier than a plain array, and the expected # of elements is low (generally 1).

Comment (mdiep) [inline ReactiveCocoa/DelegateProxy.swift]: > > It seems like we could instead create an AssociationKey from setter since it's a Selector.
>
> It could collide with others if we use it directly. In the meantime, it could use a prefixed selector, but I'd prefer to avoid a new selector.

Why? This seems like a much preferable option. The selector mangling could happen in AssociationKey so that everything else gets it for free.

Comment (andersio) [inline ReactiveCocoa/DelegateProxy.swift]: Changed in https://github.com/ReactiveCocoa/ReactiveCocoa/pull/3435/commits/78346b55202b4f688a0f846f4c6082912b7f23e7.

File: ReactiveCocoa/DelegateProxy.swift [hunk 4]
```diff
@@ -53,57 +53,64 @@ extension DelegateProxy {
...
 		setter: Selector,
-		getter: Selector,
-		_ key: StaticString = #function
+		getter: Selector
 	) -> P {
-		return _proxy(for: instance, setter: setter, getter: getter, key) as! P
+		return _proxy(for: instance, setter: setter, getter: getter) as! P
 	}
 
 	private static func _proxy(
```
Comment (mdiep) [inline ReactiveCocoa/DelegateProxy.swift]: I think we could unify the proxy could here with the proxy code in `ActionProxy`.
